fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java

```diff
@@ -213,8 +208,10 @@ private void maybeLogBuildProgress(SubspaceProvider subspaceProvider) {
             Tuple rangeStart = Arrays.equals(range.begin, START_BYTES) ? null : Tuple.fromBytes(range.begin);
             Tuple rangeEnd = Arrays.equals(range.end, END_BYTES) ? null : Tuple.fromBytes(range.end);
 
+            // range.begin, if not null, is produced by getContinuation() and should be used at continuation field
+            // range.end, if not null, is either defined by caller or as a missingRange - making a top range, therefore used at range field
             RecordCursor<FDBIndexedRecord<Message>> cursor =
-                    store.scanIndexRecords(srcIndex, IndexScanType.BY_VALUE, TupleRange.between(rangeStart, rangeEnd), range.begin, scanProperties);
+                    store.scanIndexRecords(srcIndex, IndexScanType.BY_VALUE, TupleRange.between(null, rangeEnd), range.begin, scanProperties);
```

Comment:
I guess this is somewhat better than before, but it still seems to me like this isn't really doing what the range set expects one to do. In particular, I think it will complicate parallelized builds in the future and using an index to build a non-idempotent index, where you really do want the range set to be real ranges. I think the flip side is that this might mean that we'll sometimes have to do an additional read, but I think that's preferable to making the range set contain weird things.

Reply:
A question - 
Assume that, instead of a second query, we'll ask for N+1 items in cursor and provide N as an argument to the iteration function - that will set as lastResult the N+1's item after processing N items (or null, if cursor has less than N items). 
Will that work? May the cursor have less items than the requested without reaching the records store's end? I've noticed that in `byRecords` the `respectLimit` is optional - would that mean a default limit or the whole the record store?

Reply:
Discussed a little offline, but for posterity:

> Assume that, instead of a second query, we'll ask for N+1 items in cursor and provide N as an argument to the iteration function - that will set as lastResult the N+1's item after processing N items (or null, if cursor has less than N items). Will that work?

In the general case, that can cause problems, because cursors can terminate early if they hit certain "out of bands" limits, such as the time limit or the scanned bytes limit. We don't set those (currently) in the online indexer, so we should be okay to do that "for now". We may want to make sure either the code is resilient to out of band errors or throws an error if the wrong kind of limit is reached.

> I've noticed that in byRecords the respectLimit is optional - would that mean a default limit or the whole the record store?

There's no default limit, so it would scan the whole thing. The code in question is only called with `respectLimit = false` during a call to build an index in a single transaction, so it really does need to read the entire store or it will encounter correctness problems.